staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/evaluator.go

```diff
@@ -0,0 +1,237 @@
+}
+
+// ResolveName returns a value from the activation by qualified name, or false if the name
+// could not be found.
+func (a *validationActivation) ResolveName(name string) (interface{}, bool) {
+	switch name {
+	case ObjectVarName:
+		return a.object, true
+	case OldObjectVarName:
+		return a.oldObject, true
+	case ParamsVarName:
+		return a.params, true
+	case RequestVarName:
+		return a.request, true
+	default:
+		return nil, false
+	}
+}
+
+// Parent returns the parent of the current activation, may be nil.
+// If non-nil, the parent will be searched during resolve calls.
+func (a *validationActivation) Parent() interpreter.Activation {
+	return nil
+}
+
+// Compile compiles the cel expressions defined in the CELValidations into an Evaluator
+func (c *evaluatorCompiler) Compile(validations []Validation, hasParam bool) Evaluator {
+	if len(validations) == 0 {
+		return nil
+	}
+	compilationResults := make([]CompilationResult, len(validations))
+	for i, validation := range validations {
+		compilationResults[i] = CompileValidatingPolicyExpression(validation.Expression, hasParam)
+	}
+	return NewEvaluator(compilationResults, validations)
+}
+
+// evaluator implements the Evaluator interface
+type evaluator struct {
+	compilationResults []CompilationResult
+	validations        []Validation
+}
+
+func NewEvaluator(compilationResults []CompilationResult, validations []Validation) Evaluator {
+	return &evaluator{
+		compilationResults,
+		validations,
+	}
+}
+
+func convertObjectToUnstructured(obj interface{}) (*unstructured.Unstructured, error) {
+	if obj == nil || reflect.ValueOf(obj).IsNil() {
+		return &unstructured.Unstructured{Object: nil}, nil
+	}
+	ret, err := runtime.DefaultUnstructuredConverter.ToUnstructured(obj)
+	if err != nil {
+		return nil, err
+	}
+	return &unstructured.Unstructured{Object: ret}, nil
+}
+
+func objectToResolveVal(r runtime.Object) (interface{}, error) {
+	if r == nil || reflect.ValueOf(r).IsNil() {
+		return nil, nil
+	}
+	v, err := convertObjectToUnstructured(r)
+	if err != nil {
+		return nil, err
+	}
+	return v.Object, nil
+}
+
+// Evaluate evaluates the compiled CEL expressions converting them into CELEvaluations
+// errors per evaluation are returned on the Evaluation object
+func (v *evaluator) Evaluate(versionedAttr *generic.VersionedAttributes, versionedParams runtime.Object) ([]Evaluation, error) {
+	// TODO: replace unstructured with ref.Val for CEL variables when native type support is available
+	evaluations := make([]Evaluation, len(v.compilationResults))
+	var err error
+
+	oldObjectVal, err := objectToResolveVal(versionedAttr.VersionedOldObject)
+	if err != nil {
+		return nil, err
+	}
+	objectVal, err := objectToResolveVal(versionedAttr.VersionedObject)
+	if err != nil {
+		return nil, err
+	}
+	paramsVal, err := objectToResolveVal(versionedParams)
+	if err != nil {
+		return nil, err
+	}
+
+	request := createAdmissionRequest(versionedAttr.Attributes)
```

Comment:
In the case of webhooks, we would have already constructed a v1.AdmissionReview to send to the webhook. It might be better to just pass a `*v1.AdmissionReview` to this function (`Evaluate`) instead, and leave it to the caller to construct the request.

Reply:
Are there any CEL-specific aspects to this constructor? Are we trying to reserve the right to do additional field scoping?

In any case, if we want to put it on the caller to do this conversion, we should make this, or an (equivalent?) function you mention earlier, public.

Reply:
I'm assuming you mean a third param @tallclair for the admissionrequest?

Reply:
Implemented like that